[forge] swarm network chaos

rustielin opened the pull request:
### Description

Introduce network chaos at the swarm level. New test suite `land_blocking_real` attempts to run tests against three types of network chaos, in a new experiment called `NetworkChaosTest`. In three independent steps, the experiment injects latency, limits bandwidth, and partitions the network, all while using txn-emitter to validate the network is still healthy.

Abstractly, we're introducing `SwarmChaos` with options `Delay`, `Bandwidth`, and `Partition`. Then we can do `Swarm::inject_chaos(SwarmChaos)` and `Swarm::remove_chaos(SwarmChaos)`. It's unimplemented for `LocalSwarm`, but for `K8sSwarm`, it creates/manages/deletes the NetworkChaos CRDs from Chaos Mesh: https://chaos-mesh.org/docs/simulate-network-chaos-on-kubernetes/

Also included is some boilerplate for `NodeChaos`, which is applied on individual nodes rather than the entire Swarm. One can imagine this is useful in specifying more detailed network topologies -- such as for simulating N-regions.

Caveats:
* Chaos CRDs seem to need to be applied to already running pods (otherwise are stuck in `Injecting` state infinitely / have a long backoff). This necessitates per-namespace chaos being injected at runtime, after network setup. 
* Txn-emitter needs to be tuned to account for higher latency and potentially stuck nodes (e.g. when they're split from the >2/3 supermajority)
* Chaos CRDs are applied via string template. There may be a way to specify them via https://docs.rs/kube/0.42.0/kube/derive.CustomResource.html instead

### Test Plan

Run the new test suite in Forge on the `forge-1` cluster. It's configured with 4vCPU, 8GB machines. Test fails in the final stage after 30% partition, even though the network continues to be live, because of txn-emitter workload

```
$ FORGE_RUNNER_MODE=local FORGE_TEST_SUITE=land_blocking_real ./testsuite/run_forge.sh
...
Tests Failed

failures:
    network::inject-chaos

test result: FAILED. 0 passed; 1 failed; 0 filtered out

Error: Tests Failed
{
  "metrics": [
    {
      "test_name": "network::inject-chaos:delay",
      "metric": "submitted_txn",
      "value": 17145.0
    },
    {
      "test_name": "network::inject-chaos:delay",
      "metric": "expired_txn",
      "value": 45.0
    },
    {
      "test_name": "network::inject-chaos:delay",
      "metric": "avg_tps",
      "value": 570.0
    },
    {
      "test_name": "network::inject-chaos:delay",
      "metric": "avg_latency",
      "value": 6917.0
    },
    {
      "test_name": "network::inject-chaos:delay",
      "metric": "p99_latency",
      "value": 11800.0
    },
    {
      "test_name": "network::inject-chaos:bandwidth",
      "metric": "submitted_txn",
      "value": 13365.0
    },
    {
      "test_name": "network::inject-chaos:bandwidth",
      "metric": "expired_txn",
      "value": 30.0
    },
    {
      "test_name": "network::inject-chaos:bandwidth",
      "metric": "avg_tps",
      "value": 444.0
    },
    {
      "test_name": "network::inject-chaos:bandwidth",
      "metric": "avg_latency",
      "value": 14104.0
    },
    {
      "test_name": "network::inject-chaos:bandwidth",
      "metric": "p99_latency",
      "value": 26600.0
    }
  ],
...
Injected 80ms +- 20ms with 10% correlation latency to namespace
network::inject-chaos:delay : 570 TPS, 6917 ms latency, 11800 ms p99 latency,(!) expired 45 out of 17145 txns
Limited bandwidth to 100mbps with limit 20971520 and buffer 10000 to namespace
network::inject-chaos:bandwidth : 444 TPS, 14104 ms latency, 26600 ms p99 latency,(!) expired 30 out of 13365 txns
Partitioned 30% validators in namespace
```

and see that the NetworkChaos resources are created

```
$ kubectl get networkchaos                   
NAME                                   ACTION      DURATION
forge-namespace-100mbps-bandwidth      bandwidth   
forge-namespace-30-percent-partition   partition   
forge-namespace-80ms-latency           delay       
```

Can also get to the chaos dashboard and view the network chaos 

```
# get the dashboard
$ kubectl port-forward -n chaos-mesh svc/chaos-dashboard 8333:2333
```


<!-- Please provide us with clear details for verifying that your changes work. -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/aptos-labs/aptos-core/2111)
<!-- Reviewable:end -->

perryjrandall commented:
Looking awesome just a couple questions

1) why do we run traffic generation three times can we just apply the chaos once and run traffic once?

2) can we split the abstraction of test from the application of chaos? Mutating the cluster as a side effect of a test is going to inhibit our ability to run larger more complex suites. Actually the effect abstraction from cluster test seemed pretty good

3) at least some of this code needs unit tests, it's ok to follow up in the next pr with them, also can you include the commands you run in the test plan so others can easily reproduce locally

rustielin commented:
> why do we run traffic generation three times can we just apply the chaos once and run traffic once?

Under the hood, they're different types of chaos so they need to be applied one at a time. Running traffic in between each application might tell us something about the workload 

> can we split the abstraction of test from the application of chaos? Mutating the cluster as a side effect of a test is going to inhibit our ability to run larger more complex suites. Actually the effect abstraction from cluster test seemed pretty good

I spent quite a bit of time thinking about this, and just cleaned up some of the `Chaos` abstractions (see the updated PR comment). Chaos can be applied to either `Swarm` or `Node`, and is all reversible. The abstraction of test and the application of chaos are separate, but today it's only useful to specify chaos during tests.

sitalkedia commented on testsuite/forge-cli/src/main.rs:
```diff
@@ -330,6 +330,7 @@ fn get_changelog(prev_commit: Option<&String>, upstream_commit: &str) -> String
 fn get_test_suite(suite_name: &str) -> ForgeConfig<'static> {
     match suite_name {
         "land_blocking_compat" => land_blocking_test_compat_suite(),
+        "land_blocking_real" => land_blocking_real_suite(),
```
nit: `land_blocking_real_suite` -> `land_blocking_release_suite` may be?

rustielin replied:
Good point -- I just got rid of some old tests that we haven't verified with the new setup yet, and made this the `pre_release` suite.

sitalkedia commented on testsuite/forge/src/backend/k8s/chaos/network_delay.yaml:
```diff
@@ -0,0 +1,25 @@
+kind: NetworkChaos
```
What are the thoughts on how can we have  a delay test that can inject delay as per a distribution - something like p90 delay of 1s and p50 delay of 300 ms?

rustielin replied:
I'm not sure if we can be so exact with the distribution, but with the `correlation` and `jitter` configs for `latency` chaos, we should be able to get pretty close: https://chaos-mesh.org/docs/simulate-network-chaos-on-kubernetes/#delay

sitalkedia commented on testsuite/forge/src/backend/k8s/chaos/network_bandwidth.yaml:
```diff
@@ -0,0 +1,15 @@
+apiVersion: chaos-mesh.org/v1alpha1
+metadata:
+  namespace: {namespace}
+  name: forge-namespace-{rate}mbps-bandwidth
```
Do we need a `label` selector for this as well?

rustielin replied:
ah yea we do -- good catch!
